src/id_list.c

```diff
@@ -18,6 +18,44 @@ static inline int isEof(const IdListIterator *it) {
   return !it->base.isValid;
 }
 
+typedef struct ILCriteriaTesterCtx{
+  t_docId *docIds;
+  t_offset size;
+}ILCriteriaTesterCtx;
+
+static int cmp_docids(const void *p1, const void *p2);
+
+static int IL_TextCriteria(void *ctx, t_docId id){
+  ILCriteriaTesterCtx* ctc = ctx;
+  return bsearch((void*)id, ctc->docIds, (size_t)ctc->size, sizeof(t_docId), cmp_docids) != NULL;
+}
+
+static void IL_CriteriaTesterFree(struct IndexCriteriaTester* ct){
+  ILCriteriaTesterCtx* ctc = ct->ctx;
+  rm_free(ctc->docIds);
+  rm_free(ctc);
+  rm_free(ct);
+}
+
+IndexCriteriaTester* IL_GetCriteriaTester(void *ctx){
+  IdListIterator *it = ctx;
+  ILCriteriaTesterCtx* ctc = rm_malloc(sizeof(*ctc));
+  ctc->docIds = it->docIds;
+  ctc->size = it->size;
+  it->docIds = NULL;
+  it->size = 0;
+  IndexCriteriaTester* ct = rm_malloc(sizeof(*ct));
```

Comment:
I'd rather there not be a separate "ctx" or whatever pointer. This just adds more structures and makes things confusing. Better to just use struct overlays e..g "subclassing".